src/preferences.cpp

```diff
@@ -0,0 +1,546 @@
+                valid = false;
+            }
+            if(valid)
+            ui->ddownmap->setText(QString(buffer));
+        }
+    }
+    else if(sender == ui->drightmap) {
+        if(event->type() == QEvent::KeyRelease) {
+            QKeyEvent* key_press = static_cast<QKeyEvent*>(event);
+            this->DRIGHT = key_press->nativeVirtualKey();
+            char buffer[256];
+            get_scan_code(key_press->nativeVirtualKey(), buffer, 256);
+            ui->drightmap->setText(QString(buffer));
+        }
+        else if(event->type() == QEvent::MouseButtonRelease) {
+            QMouseEvent* mouse_press = static_cast<QMouseEvent*>(event);
+            char buffer[256];
+            bool valid = true;
+            UINT button = mouse_press->button();
+            switch(button) {
+            case Qt::MouseButton::LeftButton:
+                this->DRIGHT = VK_LBUTTON;
+                get_scan_code(VK_LBUTTON, buffer, 256);
+                break;
+            case Qt::MouseButton::RightButton:
+                this->DRIGHT = VK_RBUTTON;
+                get_scan_code(VK_RBUTTON, buffer, 256);
+                break;
+            case Qt::MouseButton::MiddleButton:
+                this->DRIGHT = VK_MBUTTON;
+                get_scan_code(VK_MBUTTON, buffer, 256);
+                break;
+            default:
+                qDebug() << "Some Error Occured No Legal Mouse Button found";
+                valid = false;
+            }
+            if(valid)
+            ui->drightmap->setText(QString(buffer));
+        }
+    }
+    else if(sender == ui->dleftmap) {
+        if(event->type() == QEvent::KeyRelease) {
+            QKeyEvent* key_press = static_cast<QKeyEvent*>(event);
+            this->DLEFT = key_press->nativeVirtualKey();
+            char buffer[256];
+            get_scan_code(key_press->nativeVirtualKey(), buffer, 256);
+            ui->dleftmap->setText(QString(buffer));
+        }
+        else if(event->type() == QEvent::MouseButtonRelease) {
+            QMouseEvent* mouse_press = static_cast<QMouseEvent*>(event);
+            char buffer[256];
+            bool valid = true;
+            UINT button = mouse_press->button();
+            switch(button) {
+            case Qt::MouseButton::LeftButton:
+                this->DLEFT = VK_LBUTTON;
+                get_scan_code(VK_LBUTTON, buffer, 256);
+                break;
+            case Qt::MouseButton::RightButton:
+                this->DLEFT = VK_RBUTTON;
+                get_scan_code(VK_RBUTTON, buffer, 256);
+                break;
+            case Qt::MouseButton::MiddleButton:
+                this->DLEFT = VK_MBUTTON;
+                get_scan_code(VK_MBUTTON, buffer, 256);
+                break;
+            default:
+                qDebug() << "Some Error Occured No Legal Mouse Button found";
+                valid = false;
+            }
+            if(valid)
+            ui->dleftmap->setText(QString(buffer));
+        }
+    }
```

Comment:
Much of this code is simply duplicated with different variables.

This must change.

Reply:
I recommend creating a new component for input (like [the one discussed](https://github.com/kitswas/VirtualGamePad-PC/pull/5#issuecomment-1703814023)) and moving this logic there.